regex/src/main/scala/regex/Regex.scala

```diff
@@ -1,35 +1,70 @@
 package regex
 
 /** *****************************************************************************
-  * Regular Languages
-  *
-  * data structure definitions for regular languages
-  */
+  * Regular Languages
+  *
+  * data structure definitions for regular languages
+  */
 
-// Add your definitions here
+trait RegularLanguage
 
-/** *****************************************************************************
-  * Derivatives
-  *
-  * Fill in the function definitions below
-  */
+case object Empty extends RegularLanguage
+
+case object Epsilon extends RegularLanguage
+
+case class Character(c: Char) extends RegularLanguage
+
+case class Union(l1: RegularLanguage, l2: RegularLanguage) extends RegularLanguage
+
+case class Concat(l1: RegularLanguage, l2: RegularLanguage) extends RegularLanguage
+
+case class Star(l: RegularLanguage) extends RegularLanguage
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = lang match {
+  case Concat(Epsilon, sublang) => simplify(sublang)
+  case Concat(sublang, Epsilon) => simplify(sublang)
+  case Concat(Empty, sublang) => simplify(Empty)
+  case Concat(sublang, Empty) => simplify(Empty)
+  case Concat(sublang1, sublang2) => Concat(simplify(sublang1), simplify(sublang2))
+  case Union(Empty, sublang) => simplify(sublang)
+  case Union(sublang, Empty) => simplify(sublang)
+  case Union(sublang1, sublang2) => Union(simplify(sublang1), simplify(sublang2))
+  case Star(Epsilon) => Epsilon
+  case Star(Empty) => Empty
+  case Star(sublang) => Star(simplify(sublang))
+  case _ => lang
+}
```

Comment:
I liked the use of "sublang" here, I referred to these as lang1 and lang2. However, it's a bit tough to see 10 lines of cases, instead I'd suggest using match Concat(sublang1, sublang2) then making each option a case, that way the Concat, Union, and Star cases are all separate.